app/controllers/businessactivities/WhoIsYourAccountantIsUkController.scala

```diff
@@ -0,0 +1,93 @@
+
+package controllers.businessactivities
+
+import com.google.inject.Inject
+import connectors.DataCacheConnector
+import controllers.DefaultBaseController
+import forms.{EmptyForm, Form2, InvalidForm, ValidForm}
+import models.businessactivities.{BusinessActivities, WhoIsYourAccountantIsUk}
+import services.AutoCompleteService
+import uk.gov.hmrc.play.frontend.auth.connectors.AuthConnector
+import utils.AuthAction
+
+class WhoIsYourAccountantIsUkController @Inject()(val dataCacheConnector: DataCacheConnector,
+                                                  val autoCompleteService: AutoCompleteService,
+                                                  val authAction: AuthAction
+                                              )extends DefaultBaseController {
+
+  def get(edit: Boolean = false) = authAction.async {
+    implicit request =>
+      dataCacheConnector.fetch[BusinessActivities](request.credId, BusinessActivities.key) map {
+        response =>
+          val form = getForm(response)
+          val name = getName(response)
+          Ok(views.html.businessactivities.who_is_your_accountant_is_uk_address(form, edit, name))
+      }
+  }
+
+  def post(edit : Boolean = false) = authAction.async {
+    implicit request =>
+      Form2[WhoIsYourAccountantIsUk](request.body) match {
+        case f: InvalidForm =>
+          dataCacheConnector.fetch[BusinessActivities](request.credId, BusinessActivities.key) map {
+            response => BadRequest(views.html.businessactivities.who_is_your_accountant_is_uk_address(f, edit, getName(response)))
+          }
+        case ValidForm(_, data) =>
+          for {
+            businessActivity <- dataCacheConnector.fetch[BusinessActivities](request.credId, BusinessActivities.key)
+            _ <- dataCacheConnector.save[BusinessActivities](request.credId, BusinessActivities.key, updateModel(businessActivity, data))
+          } yield if (data.isUk) {
+            Redirect(routes.WhoIsYourAccountantUkAddressController.get(edit))
+          } else {
+            Redirect(routes.WhoIsYourAccountantNonUkAddressController.get(edit))
+          }
+      }
+  }
+
+  private def getForm(response: Option[BusinessActivities]): Form2[_] = {
+    val isUkForm = for {
+      businessActivities <- response
+      isUk <- businessActivities.whoIsYourAccountant.flatMap(acc => acc.isUk)
+    } yield {
+      Form2[WhoIsYourAccountantIsUk](isUk)
+    }
+
+    isUkForm.getOrElse {
+      val inferredIsUkForm = for {
+        businessActivities <- response
+        isUk <- businessActivities.whoIsYourAccountant.flatMap(acc => acc.address).map(address => address.isUk)
+      } yield {
+        Form2[WhoIsYourAccountantIsUk](WhoIsYourAccountantIsUk(isUk))
+      }
+      inferredIsUkForm.getOrElse(EmptyForm)
+    }
+  }
+
+  private def getName(response: Option[BusinessActivities]): String = response
+    .flatMap(ba => ba.whoIsYourAccountant).flatMap(acc => acc.names).map(acc => acc.accountantsName).getOrElse("")
```

Comment:
There are a lot of maps and flat maps in one line here. Can this be refactored out in separate descriptive defs? 

I.e.

```
 private def getAccountant(businessActivities: Option[BusinessActivities]) = {
    businessActivities.flatMap {
      ba => ba.whoIsYourAccountant
    }
  }

  private def getNames(businessActivities: Option[BusinessActivities]) = {
    getAccountant(businessActivities).flatMap{
      acc => acc.names
    }
  }

  private def getName(response: Option[BusinessActivities]): String =
    getNames(response).map(acc => acc.accountantsName).getOrElse("")
```

This could then be put into a a shared resource and reused in the other controllers?

Reply:
Changed to a for-comprehension and moved to a superclass